README.md
<!-- PROJECT LOGO -->
<br />
<p align="center">
  <a href="https://github.com/ocular-d/vagrant-vscode">
    <img src="docs/assets/ocular-d-logo.png" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">Vagrant VSCode</h3>

  <p align="center">
    Visual Studio Code Demo For Technical Writing
    <br />
    <a href="https://github.com/ocular-d/vagrant-vscode"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/ocular-d/Best-vagrant-vscode/issues">Report Bug</a>
    ·
    <a href="https://github.com/ocular-d/vagrant-vscode/issues">Request Feature</a>
  </p>
</p>

<!-- TABLE OF CONTENTS -->
## Table of Contents

- [About the Project](#about)
  - [Requirements](#requirements)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)
- [Credits](#credits)

<!-- ABOUT THE PROJECT -->
## About

This is a demo of how to use [Visual Studio Code](https://code.visualstudio.com/ "Link to website") for technical (API) writing.

VSCode (Visual Studio Code) is running on top of an [Ubuntu 18.04 LTS Desktop](https://ubuntu.com/download/desktop "Website of Ubuntu Desktop") configured with [extensions](https://code.visualstudio.com/docs/editor/extension-gallery "Link to website with VSCode extensions") for writing in [Markdown](https://en.wikipedia.org/wiki/Markdown "Link to wikipedia").

### Requirements

- [Vagrant](https://www.vagrantup.com/ "Link to website of Vagrant")
- [Git](https://git-scm.com/ "Link to website of Git")
- [VirtualBox](https://www.virtualbox.org/ "Link to VirtualBox")

## Usage

### Getting started

Open you terminal and run:

```shell
vagrant init ocular-d/vscode-demo
```

| **Note:** Since this box is based on a full Ubuntu desktop, the image is large (around 2.G).

### Building locally

First make sure that you have all [requirements](#requirements) installed.

Now open a terminal and *clone* the repository:

```shell
git clone https://github.com/ocular-d/vagrant-vscode.git
```

Change into that directory:

```shell
cd vagrant-vscode
```

Start the setup process by starting vagrant:

```shell
vagrant up
```

![vagrant up 1](docs/assets/vagrant-up1.png "vagrant up 1")

The first time running `vagrant up` will install and setup all required vagrant plugins.

After the first run is finished, please run `vagrant up` for a second time.

![vagrant up 2](docs/assets/vagrant-up2.png "vagrant up 2")

Now you will see the boot process starting.

Depending on your bandwidth, this can take some minutes.

After the build and setup process is finished, you can login in with the following credentials:

**user:** *vagrant*

**password:** *vagrant*

## Contributing

We are a community effort, and everybody is most welcome to participate!

Be it filing bugs, formulating enhancements, creating pull requests,
or any other means of contribution, we encourage contributions from everyone.

## License

Distributed under the [GPLv2](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html "Link to license") license.

## Credits

- [Pronovix](https://pronovix.com/ "Link to Pronovix website")


CHANGELOG.md
# Changelog

## 0.0.2

- Update base box
- Add custom Vagrantfile during box creation
- Switch desktop install to `--no-install-recommends`

## 0.0.1

- Initial release


setup.sh
#!/bin/bash -eux

set -euo pipefail

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
#COL_RED=$ESC_SEQ"31;01m"
#COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"

# Functions
# Update
extensions() {
    echo
    echo -en "${COL_YELLOW} ==> Installing VSCode extensions ${COL_RESET}\n"
    code --install-extension pnp.polacode
    code --install-extension bierner.markdown-emoji
    code --install-extension AlanWalk.markdown-toc
    code --install-extension CoenraadS.bracket-pair-colorizer
    code --install-extension streetsidesoftware.code-spell-checker
    code --install-extension ms-vscode.wordcount
    code --install-extension ybaumes.highlight-trailing-white-spaces
    code --install-extension myh.preview-vscode
    code --install-extension mrmlnc.vscode-remark
    code --install-extension johnpapa.read-time
    code --install-extension vincaslt.highlight-matching-tag
    code --install-extension PKief.material-icon-theme
    code --install-extension Hyzeta.vscode-theme-github-light
    code --install-extension eamodio.gitlens
    code --install-extension drewbourne.vscode-remark-lint
    code --install-extension testthedocs.vale
}

npm_modules() {
    npm install remark-cli remark-preset-lint-recommended
    npm install remark-preset-lint-consistent remark-validate-links
    npm install remark-preset-lint-markdown-style-guide
}

update_bashrc() {
    echo 'export PATH="$PATH:/home/vagrant/node_modules/.bin"' >> ~/.bashrc
}

update_vscode_settings() {
    wget https://gist.githubusercontent.com/ocular-d/cda72372a8168f0711700d417fa8a13e/raw/5f23d12b395dd62d271c02e5ee3af26a000c06c7/settings.json -O /$HOME/.config/Code/User/settings.json
}

# Run functions
extensions
npm_modules
update_bashrc
update_vscode_settings


script.sh
#!/bin/bash -eux

set -euo pipefail

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
#COL_RED=$ESC_SEQ"31;01m"
#COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"

# Functions
# Update
update() {
    echo
    echo -en "${COL_YELLOW} ==> Updating sources ${COL_RESET}\n"
    apt update -y -qq
}


# Upgrade
upgrade() {
    echo -en "${COL_YELLOW} ==> Upgrading packages to latest versions ${COL_RESET}\n"
    apt upgrade -y -qq
}

#The script is used for graphic environment
vb-guest() {
    echo -en "${COL_YELLOW} ==> Installing VirtualBox guest extension ${COL_RESET}\n"
    if grep -q VBOX /proc/scsi/scsi; then
        echo "*** Installing virtualbox-guest-x11"

    export DEBIAN_FRONTEND="noninteractive"
    apt install -y -qq --no-install-recommends virtualbox-guest-x11
    fi
}

# Install Ubuntu Desktop
desktop() {
    echo -en "${COL_YELLOW} ==> Installing Ubuntu desktop ${COL_RESET}\n"
    apt install -y -qq ubuntu-desktop
}

# Use snap to install Code
vscode() {
    echo -en "${COL_YELLOW} ==> Installing Visual Studio Code ${COL_RESET}\n"
    snap install --classic code
}

update
upgrade
vb-guest
desktop
vscode


install.sh
#!/bin/bash -eux

set -euo pipefail

# Colors
ESC_SEQ="\x1b["
COL_RESET=$ESC_SEQ"39;49;00m"
#COL_RED=$ESC_SEQ"31;01m"
#COL_GREEN=$ESC_SEQ"32;01m"
COL_YELLOW=$ESC_SEQ"33;01m"

# Functions
# Update
update() {
    echo
    echo -en "${COL_YELLOW} ==> Updating sources ${COL_RESET}\n"
    apt update -y -qq
}


# Upgrade
upgrade() {
    echo -en "${COL_YELLOW} ==> Upgrading packages to latest versions ${COL_RESET}\n"
    apt upgrade -y -qq
}

# The script is used for graphic environment
#vb-guest() {
#    echo -en "${COL_YELLOW} ==> Installing VirtualBox guest extension ${COL_RESET}\n"
#    if grep -q VBOX /proc/scsi/scsi; then
#        echo "*** Installing virtualbox-guest-x11"
#
#    export DEBIAN_FRONTEND="noninteractive"
#    apt install -y -qq --no-install-recommends virtualbox-guest-x11
#    fi
#}

# Install Ubuntu Desktop
desktop() {
    echo -en "${COL_YELLOW} ==> Installing Ubuntu desktop ${COL_RESET}\n"
    apt install -y --no-install-recommends ubuntu-desktop fonts-firacode
}

# Use snap to install Visual Studio Code
vscode() {
    echo -en "${COL_YELLOW} ==> Installing Visual Studio Code ${COL_RESET}\n"
    snap install --classic code
}

npm_install() {
   snap install node --channel=10/stable --classic
}

vale_install() {
    curl -sfL https://install.goreleaser.com/github.com/ValeLint/vale.sh | sudo bash -s -- -b /usr/local/bin
}

# Run functions
update
upgrade
#vb-guest
desktop
vscode
npm_install
vale_install
